README.md
# REMO-IMOVE_SPITFIRE

autosubmit/ChangeLog
2017-04-04 Lars Buntemeyer <lars.buntemeyer@hzg.de>
	Added all dynamics modules and two tracer fields.
	TRAC1 and TRAC2 and two source fields TR1SOURCE and TR2SOURCE.
	progexp was converted to fortran90 standard.

	Added:

	* mo_dynamics.f90
	* mo_memory_dynamics.f90
	* progexp.f90

2017-03-27 Lars Buntemeyer <lars.buntemeyer@hzg.de>
	Added all memory modules for dynamica allocation and
	new netcdf IO. For more details, please look at the
	user guide and README. param.h has been replaced by
	mo_param.

	Added:

	* mo_*
	* control.f90
	* init_remo.f90 timeloop.f90 finish_remo.f90
	* io_initial.f90 io_restart.f90

2017-03-09 Lars Buntemeyer <lars.buntemeyer@hzg.de>
	Added REMO documentation as git submodule

	Modified:

	* ./doc

2016-11-11 Lars Buntemeyer <lars.buntemeyer@hzg.de>
	Dynamic memory allocation in mo_sumfel.f:
	Fields for accumulation of min/max and mean values are
	dynamically allocated during runtime so that only memory
	Dynamic memory allocation in mo_sumfel.f:
	Fields for accumulation of min/max and mean values are
	dynamically allocated during runtime so that only memory
	for requested fields is allocated. Furthermore, min/max and mean
	computations are now done in parallel before communication, 
	allocation of global fields is vastly reduced for the mpi master.

	Modified:

	* mo_sumfel.f
	* writegb.f
	* tagmit.f
	* monmit.f

2016-11-10  Kevin Sieck <kevin.sieck@hzg.de>

	Renamed some autosubmit scripts to achieve a better naming.
	Added postprocessing script.

	Modified:

	* autosubmit/conf/jobs_xxxx.conf
        * autosubmit/conf/proj_xxxx.conf
	* autosubmit/jobs/member_setup.py
	* autosubmit/jobs/remote_setup.sh

        Deleted:

	* autosubmit/jobs/local_setup.sh

        Added:

	* autosubmit/jobs/model_setup.sh
        * autosubmit/jobs/postprocess.sh

2016-11-09  Kevin Sieck <kevin.sieck@hzg.de>

	Included boundary data download script for autosubmit.
	Many little changes to autosubmit scripts.

	Added:

	* autosubmit/jobs/ret_forcing.sh

	Modified:

	* autosubmit/conf/proj_xxxx.conf
        * autosubmit/jobs/local_setup.sh
        * autosubmit/jobs/member_setup.py
        * autosubmit/jobs/remote_setup.sh
        * autosubmit/jobs/test_remo.sh

2016-10-25  Kevin Sieck <kevin.sieck@hzg.de>

	Added a new directory (autosubmit/..) that contains jobs and config
	files that define a basic workflow to run REMO using autosubmit.

	For now autosubmit needs some changes in the slurmplatform file. The
	account and partition need to be added to the slurm header in order
	to run on DKRZ computers.

	Added files:

	Needed for autosubmit:
	* autosubmit/conf/proj_xxxx.conf
	* autosubmit/conf/jobs_xxxx.conf

	Basic configuration files for REMO Namelists:
	* autosubmit/conf/eur011.conf
	* autosubmit/conf/eur044.conf

	Autosubmit job files:
	* autosubmit/jobs/local_setup.sh
	* autosubmit/jobs/remote_setup.sh
	* autosubmit/jobs/member_setup.py
	* autosubmit/jobs/test_remo.sh

2016-07-20 Thomas Raub <thomas.raub@hzg.de>
	Adjustments to support more than 49 levels for the netcdf version.

	Modified:

	* ecaprepnc.f90
	* ecrprepnc.f90
	* mo_netcdf.f90
	* writenc.f90

2016-07-19 Thomas Raub <thomas.raub@hzg.de>
	Merged model_top branch into remo2011_nh_debug

2016-06-28  Kevin Sieck <kevin.sieck@hzg.de>

	Fixed a small bug which lead to chrashes when model top was not 0.0Pa.

	* CODE/progexp.f
	* CODE/ectied.f

2016-06-24  Kevin Sieck <kevin.sieck@hzg.de>

	Implemented a model top which can be different from 0.0 Pa.

	Modified:

	All routines that computed the pressure with ak, bk

	Added:

	* CODE/getp.f
	* CODE/piber.f

2016-06-10  Lars Buntemeyer <lars.buntemeyer@hzg.de>

	Added netcdf support.
	Moved all long-term data from ec4org.f to
	the module EM_Data and EC4_Data.

	Added:

	* CODE/mo_io.f90
	* CODE/mo_netcdf.f90
	* CODE/ecaprepnc.f90
	* CODE/ecrprepnc.f90
	* CODE/makefile.io
	* CODE/makefile.netcdf
	* CODE/EM_Data.f90
	* CODE/EC4_Data.f90

2016-06-06  Kevin Sieck <kevin.sieck@hzg.de>

	Introduced rayleigh damping for idealized hill tests with
	switch LRAYDAMP. Default is .FALSE.

	Modified:

	* CODE/progec4.f
	* CODE/medea.f
	* CBS/mo_comdyn.f

	Added:

	* CODE/rayleigh.f

2016-05-10  Kevin Sieck <kevin.sieck@hzg.de>

	Changed the non-hydrostatic switch to calculate W also for the
	hydrostatic case.

	Modified:

	* CODE/progec4.f
	* CODE/cdzdt.f

2016-05-09  Kevin Sieck <kevin.sieck@hzg.de>

	Added makefile for eddy.

2016-05-09  Kevin Sieck <kevin.sieck@hzg.de>

	Fixed bug in call to progexp.
	Extended the non-hydrostatic correction calculations to the boundaries
	(JAHCOMP instead of JAH, etc.).
	Implemented the divergence damping analog to WRF.
	Introduced WR as boundary field for W.

	Modified:

	* CBS/mo_comdyn.f
	* CODE/cdwdt.f
        * CODE/cdwdt2.f
        * CODE/cdzdt.f
        * CODE/ec4inmi.f
        * CODE/ec4org.f
        * CODE/ecaprep.f
        * CODE/ecrandup.f
        * CODE/final.f
        * CODE/gkonst.f
        * CODE/medea.f
        * CODE/progec4.f
        * CODE/progexp.f
        * jobs/test_remo_2d.ksh

2016-04-04  Thomas Raub <thomas.raub@hzg.de>

	Fixed bugs in adv.f (and adv_kev.f) which led to wrong sign and
	amplitude of horizontal advection of height and vertical velocities.

	Modified:

	* CODE/adv.f
	* CODE/adv_kev.f
	* CODE/cdwdt.f
	* CODE/cdzdt.f
	* CODE/progec4.f

2016-02-25  Kevin Sieck <kevin.sieck@hzg.de>

	Implemented the the calculation of the 10m wind for lowest layers that
	have a half smaller than 10m. In this case the 10m wind is interpolated
	(linear) between the two lowest model layers.

	* CODE/vdiff.f

2016-01-20  Kevin Sieck <kevin.sieck@hzg.de>

	Moved common parts of exchgcoeff_land.f and exchgcoeff_water.f to
	subroutines.

	Added:

	* CODE/pbl_extension.f
	* CODE/richardson.f

	Modified:

	* CODE/exchgcoeff_land.f
	* CODE/exchgcoeff_water.f
	* CODE/vdiff.f

2016-01-20  Kevin Sieck <kevin.sieck@hzg.de>

	Moved the calculation of surface specific humidity to subroutine.

	Added:

	* CODE/surfhum.f

	Changed:

	* CODE/vdiff.f

autosubmit/jobs/test_remo.sh
set -ex

# Set modelrun command depending on machine
#
if [ '%CURRENT_ARCH%' == 'LOCAL' ]
then
    function modelrun {
    time mpirun -np $NCPU ${PFL}/rremo < INPUT
    }
elif [ '%CURRENT_ARCH%' == 'dkrz' ]
then
    ulimit -s unlimited
    function modelrun {
    time srun ${PFL}/rremo < INPUT
    }
else
    echo "%CURRENT_ARCH% invalid MACHINE!"
    exit
fi

# model binary
MODELDIR=%HPCROOTDIR%/REMO_MPI # Model location
PFL=${MODELDIR}/libs

#
# Set beginning and end of simulation (works only on full day basis)
#
KSA=$(( %PREV% * 24 ))
KSE=$(( %Chunk_End_IN_DAYS% * 24 ))
#
NCPU=%NUMTASK%

#
# Number of failed simulations
#
DT=%FAIL_COUNT%
date
#
set -ex
#
# Namelist substitution
#
python -c "from string import Template
txt = Template(file('${MODELDIR}/autosubmit/INPUT_%MEMBER%_dt${DT}').read()).safe_substitute(NHANF=${KSA},NHENDE='${KSE}',NHFORA='${KSE}')
file('%HPCROOTDIR%/%MEMBER%/INPUT','w').write(txt)"
#
# model run
#
cd %HPCROOTDIR%/%MEMBER%
modelrun
#


autosubmit/jobs/putscript.py
# This skript is meant to be used to move the REMO-result files
# to the archive while performing a REMO-run.
# It moves files to a tar-file and then moves the tar-file to
# the archive-server.
#
# It is based on putscript.py from the runnigRemo directory of the old SVN
# except that it does not take any arguments, but the information is passed
# directly via autosubmit variable substitution. Some minor changes have been
# implemented
# 
##
# Notes:
# 
# If the directory in the archive already exists,
# it is renamed from e.g. 1989 to 1989_backup
# if also the latter directory exists, the program
# raises an error.
#

#import argparse
import os
import logging
import pickle
import subprocess
import tempfile

from PyRemo.Ftp import check_ftp_path, make_ftp_path, upload_file
from PyRemo.Ftp import rename_ftp_path
from PyRemo.ShellTools import check_make_path
from PyRemo.ShellTools import pack_tar_archive, extract_tar_archive
from PyRemo.ScriptTools import log_check_in
#
# check the parameters
debug = False

logging.basicConfig(level=logging.INFO,
                    format='%(asctime)s %(levelname)-8s %(message)s',
                    datefmt='%d %b %Y %H:%M:%S')

#
#parser = argparse.ArgumentParser(description='Moves REMO output to the archive.',
#        epilog='In case of problems, contact: claas.teichmann@hzg.de')
#parser.add_argument('proj_id', metavar='PROJID', type=str,
#        help='This determines the project directory in the archive.')
#parser.add_argument('expstr', metavar='EXPSTR', type=str,
#        help='The experiment string, e.g. 019400')
#parser.add_argument('year', metavar='YEAR', type=int,
#        help='The year to put to the archive, e.g. 1977.')
#parser.add_argument('--user', metavar='USER', type=str, nargs='?',
#        help='user-number for path definition, e.g., m222026',
#        default='')
#parser.add_argument('--pfiles', action='store_true',
#        help='Also uploading a tar containing p-files.')
#parser.add_argument('--repacking-yearly-efiles', action='store_true',
#        help='Repacking the monthly e-tar files to yearly e-files for each code.')
#parser.add_argument('--overwrite', action='store_true',
#        help='Overwrite existing files in the archive.')
#parser.add_argument('--wrkdir', metavar='WRKDIR', type=str, nargs='?',
#        help='Work directory for output data.')
#args = parser.parse_args()
#print args.proj_id
#print args.expstr
#print args.year
#print args.user

remo_user = '%REMO_USER%'
remo_exp = '%REMO_EXP%'

dic = {}
dic['proj_id'] = '%HPCPROJ%' #args.proj_id
if %REMO_EXP_STR_STYLE% == 1:
    dic['exp_str'] = '{0}'.format(remo_exp)
elif %REMO_EXP_STR_STYLE% == 2:
    dic['exp_str'] = '{0}{1}'.format(remo_user, remo_exp) #args.expstr
else:
    exit (1)
dic['year'] = '%Chunk_START_YEAR%' #args.year
dic['user'] = '%HPCUSER%' #args.user
dic['wrkdir'] = '%WORK_DIR%' #args.wrkdir
dic['scrdir'] = '%SCRATCH_DIR%'
year = dic['year']

do_pfiles = False #args.pfiles
do_repack_yearly_efiles = False #args.repacking_yearly_efiles
overwrite = True #args.overwrite

#
# paths
#
source_dir = "{wrkdir}/res_{exp_str}/{year}".format(**dic)
log_path = '{0}/put_log'.format(source_dir)
check_make_path(log_path)
log_file = os.path.join(log_path, 'put_{exp_str}_{year}.pickle'.format(**dic))
# path in archive accessible via pftp
dest_dir_base = "/hpss/arch/{proj_id}/{user}/exp{exp_str}".format(**dic)
dest_dir      = os.path.join(dest_dir_base, "{year}".format(**dic))
# tmp_base = os.path.join(os.getenv('SCRATCH'), 'put_tmp')
tmp_base = os.path.join("{scrdir}".format(**dic),
                        "put_tmp{exp_str}".format(**dic))
check_make_path(tmp_base)

# check the work already done

if os.path.isfile(log_file):
    f = open(log_file, 'r')
    log_dic = pickle.load(f)
    f.close()
else:
    log_dic = {}

if 'tmp_dir' in log_dic.keys():
    tmp_dir = log_dic['tmp_dir'][0]
else:
    tmp_dir = tempfile.mkdtemp(dir=tmp_base, prefix='{year}_'.format(**dic))
    log_check_in(log_file, log_dic, 'tmp_dir', tmp_dir)

# server
server = 'tape'

log_check_in(log_file, log_dic, year, 'starting')

# defining convenience funktions that could go to PyRemo
# or/and be more generalized. E.g. function call as a parameter.
def pack_tar_archive_log(source_dir, tmp_dir, tar_file_name, file_names,
        log_file, log_dic, year, task, do_remove=True):
    if task in log_dic[year]:
        logging.warning('Already done: SKIPPING {0}'.format(task))
    else:
        logging.info('Processing: {0}'.format(task))
        pack_tar_archive(source_dir, tmp_dir, tar_file_name, file_names,
                         do_remove=do_remove)
        log_check_in(log_file, log_dic, year, task)

def upload_file_log(server, tmp_dir, tar_file_name, dest_dir,
        log_file, log_dic, year, task):
    if task in log_dic[year]:
        logging.warning('Already done: SKIPPING {0}'.format(task))
    else:
        logging.info('Processing: {0}'.format(task))
        upload_file(server, tmp_dir, tar_file_name, dest_dir,
                    remove_source=True)
        log_check_in(log_file, log_dic, year, task)


# packing and uploading mns-files
file_names = []
tar_file_name = 'e{exp_str}mns{year}.tar'.format(**dic)
for dic['mon'] in range(1,13):
    for file_type in 'n':
        dic['type'] = file_type
        file_names.append('e{exp_str}{type}{year}{mon:02d}.tar'.format(**dic))
    for file_type in 'ms':
        dic['type'] = file_type
        file_names.append('e{exp_str}{type}{year}{mon:02d}'.format(**dic))

task = 'pack_tar_archive nms-files'
pack_tar_archive_log(source_dir, tmp_dir, tar_file_name, file_names,
                     log_file, log_dic, year, task, do_remove=False)
#pack_tar_archive(source_dir, tmp_dir, tar_file_name, file_names,
#                 do_remove=False)

path_exists = check_ftp_path(server, dest_dir_base)
if not path_exists:
    make_ftp_path(server, dest_dir_base)

task = 'check_ftp_directory dest_dir'
if task in log_dic[year]:
    logging.warning('Already done: SKIPPING {0}'.format(task))
else:
    logging.info('Processing: {0}'.format(task))
    path_exists = check_ftp_path(server, dest_dir)
    if path_exists:
        if overwrite:
            pass
        else:
            new_dest_dir = dest_dir + '_backup'
            rename_ftp_path(server, dest_dir, new_dest_dir)
            dest_dir = new_dest_dir
    else:
        make_ftp_path(server, dest_dir)

    log_check_in(log_file, log_dic, year, task)

task = 'upload of nms-tar'
upload_file_log(server, tmp_dir, tar_file_name, dest_dir,
                log_file, log_dic, year, task)
#upload_file(server, tmp_dir, tar_file_name, dest_dir,
#            remove_source=True)

# packing and uploading f-files
file_names = []
tar_file_name = 'e{exp_str}f{year}.tar'.format(**dic)
dic['type'] = 'f'
for dic['mon'] in range(1,13):
    file_names.append('e{exp_str}{type}{year}{mon:02d}.tar'.format(**dic))

task = 'pack_tar_archive f-files'
pack_tar_archive_log(source_dir, tmp_dir, tar_file_name, file_names,
                     log_file, log_dic, year, task)
#pack_tar_archive(source_dir, tmp_dir, tar_file_name, file_names)

task = 'upload of f-tar'
upload_file_log(server, tmp_dir, tar_file_name, dest_dir,
                log_file, log_dic, year, task)
#upload_file(server, tmp_dir, tar_file_name, dest_dir,
#            remove_source=True)

# uploading t-files
file_names = []
dic['type'] = 't'
for dic['mon'] in range(1,13):
    akt_file_name = 'e{exp_str}{type}{year}{mon:02d}.tar'.format(**dic)
    task = 'upload of t-tar for month {mon}'.format(**dic)
    upload_file_log(server, source_dir, akt_file_name, dest_dir,
                    log_file, log_dic, year, task)
#    upload_file(server, tmp_dir, tar_file_name, dest_dir,
#                remove_source=True)

# uploading e-files
dic['type'] = 'e'
if do_repack_yearly_efiles:
    if 'e_files_tmp_dir' in log_dic.keys():
        e_files_tmp_dir = log_dic['e_files_tmp_dir'][0]
    else:
        e_files_tmp_dir = tempfile.mkdtemp(dir=tmp_dir, prefix='efiles_')
        log_check_in(log_file, log_dic, 'e_files_tmp_dir', e_files_tmp_dir)
    for dic['mon'] in range(1,13):
        akt_file_name = 'e{exp_str}{type}{year}{mon:02d}.tar'.format(**dic)
        task = 'extract e-file for month {0}'.format(dic['mon'])
        if task in log_dic[year]:
            logging.warning('Already done: SKIPPING {0}'.format(task))
        else:
            logging.info('Processing: {0}'.format(task))
            extract_tar_archive(source_dir, e_files_tmp_dir, akt_file_name,
                                do_remove=True)
            log_check_in(log_file, log_dic, year, task)
    file_list = os.listdir(e_files_tmp_dir)
    code_list = set(map(lambda w: w.split('_')[1], file_list))
    logging.info('Processing codes: {0}'.format(code_list))
    for code in code_list:
        dic['code'] = code
        logging.info('repacking e-files: Processing code {code}'.format(**dic))
        file_names = []
        for dic['mon'] in range(1,13):
            file_names.append('e{exp_str}{type}_{code}_{year}{mon:02d}'.format(**dic))
        tar_file_name = 'e{exp_str}{type}_{code}_{year}{mon:02d}.tar'.format(**dic)
        task = 'pack new tar file for code: {0}'.format(code)
        pack_tar_archive_log(e_files_tmp_dir, tmp_dir, tar_file_name,
                             file_names, log_file, log_dic, year, task)
#        pack_tar_archive(e_files_tmp_dir, tmp_dir, tar_file_name, file_names)
        task = 'upload of e-tar for code: {0}'.format(code)
        upload_file_log(server, tmp_dir, tar_file_name, dest_dir,
                        log_file, log_dic, year, task)
#        upload_file(server, tmp_dir, tar_file_name, dest_dir,
#                    remove_source=True)
    os.rmdir(e_files_tmp_dir)
else:
    file_names = []
    for dic['mon'] in range(1,13):
        akt_file_name = 'e{exp_str}{type}{year}{mon:02d}.tar'.format(**dic)
        task = 'upload e-tar for month: {0}'.format(dic['mon'])
        upload_file_log(server, source_dir, akt_file_name, dest_dir,
                        log_file, log_dic, year, task)
#        upload_file(server, tmp_dir, tar_file_name, dest_dir,
#                    remove_source=True)

# uploading p-tar-file
if do_pfiles:
    dic['type'] = 'p'
    akt_file_name = 'e{exp_str}{type}{year}{mon:02d}.tar'.format(**dic)
    task = 'upload of p-tar'
    upload_file_log(server, source_dir, akt_file_name, dest_dir,
                    log_file, log_dic, year, task)
#    upload_file(server, tmp_dir, tar_file_name, dest_dir, remove_source=True)

os.rmdir(tmp_dir)


autosubmit/jobs/postprocess.sh
set -ex
#
# Moves simulated data from scratch to a different location to keep
# simulation directories clean
#
date
#
# autosubmit variables
UE=%REMO_USER%
EE=%REMO_EXP%
Y=%Chunk_START_YEAR%
M=%Chunk_START_MONTH%
YNEU=%Chunk_END_YEAR%
MNEU=%Chunk_END_MONTH%
DNEU=%Chunk_END_DAY%
HNEU=%Chunk_END_HOUR%

if [ %REMO_EXP_STR_STYLE% = 1 ]
then
    EXP_STR=$EE
elif [ %REMO_EXP_STR_STYLE% = 2 ]
then
    EXP_STR=$UE$EE
else
    exit 1
fi
# Set paths
# Paths and directories have to be the same as in member_setup.sh !!!
SIM_BASE_PATH=%HPCROOTDIR%/%MEMBER%
PFADRES=%WORK_DIR%/res_$EXP_STR
if [ '%DATA_DIRS%' = 'default' ]
then
    DATE=xe
    DATF=xf
    DATM=xm
    DATN=xn
    DATT=xt
else
    DATE=%PATH_E%
    DATF=%PATH_F%
    DATM=%PATH_M%
    DATN=%PATH_N%
    DATT=%PATH_T%
fi

TYPE=e
TYPF=f
TYPG=g
TYPM=m
TYPN=n
TYPT=t
TYPS=s
ART=.tar
#
if [[ "%Chunk_FIRST%" = "TRUE" || $M = '01' ]]
then
    mkdir -p ${PFADRES}/${Y}
fi
#
# COPY/MOVE DATA
#
# *** Restart
cd ${SIM_BASE_PATH}/${DATF}
FFILE=${TYPE}${UE}${EE}${TYPF}${YNEU}${MNEU}${DNEU}${HNEU}
GFILE=${TYPE}${UE}${EE}${TYPG}${YNEU}${MNEU}${DNEU}${HNEU}
TARFILE=${TYPE}${UE}${EE}${TYPF}${Y}${M}${ART}
SAVEFILE=${PFADRES}/${Y}/$TARFILE
if [ ! -s $SAVEFILE ]
then
    tar cf $TARFILE $FFILE $GFILE
    mv $TARFILE $SAVEFILE
fi
#
# *** Monthly files
cd ${SIM_BASE_PATH}/${DATM}
MFILE=${TYPE}${UE}${EE}${TYPM}${Y}${M}
SFILE=${TYPE}${UE}${EE}${TYPS}${Y}${M}
for MYFILE in $MFILE $SFILE
do
    SAVEFILE=${PFADRES}/${Y}/$MYFILE
    if [ ! -s $SAVEFILE ]
    then
	mv $MYFILE $SAVEFILE
    fi
done
#
# *** 3d atmospheric files
cd ${SIM_BASE_PATH}/${DATT}
TFILES=${TYPE}${UE}${EE}${TYPT}${Y}${M}
TARFILE=${TYPE}${UE}${EE}${TYPT}${Y}${M}${ART}
SAVEFILE=${PFADRES}/${Y}/$TARFILE
if [ ! -s $SAVEFILE ]
then
    tar cf $TARFILE ${TFILES}????
    mv $TARFILE $SAVEFILE
    rm ${TFILES}????
fi
#
# *** 2d files, e.g. surface etc. 
cd ${SIM_BASE_PATH}/${DATE}
EFILES=${TYPE}${UE}${EE}${TYPE}
TARFILE=${TYPE}${UE}${EE}${TYPE}${Y}${M}${ART}
SAVEFILE=${PFADRES}/${Y}/$TARFILE
if [ ! -s $SAVEFILE ]
then
    tar cf $TARFILE ${EFILES}_c???_${Y}${M} 
    mv $TARFILE $SAVEFILE
    rm ${EFILES}_c???_${Y}${M}
fi
#
# *** Daily files
cd ${SIM_BASE_PATH}/${DATN}
NFILES=${TYPE}${UE}${EE}${TYPN}
TARFILE=${TYPE}${UE}${EE}${TYPN}${Y}${M}${ART}
SAVEFILE=${PFADRES}/${Y}/$TARFILE
if [ ! -s $SAVEFILE ]
then
    tar cf $TARFILE ${NFILES}_c???_${Y}${M} 
    mv $TARFILE $SAVEFILE
    rm ${NFILES}_c???_${Y}${M}
fi
#
#
date


autosubmit/jobs/replace_codes.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# replaces parameters in an a-file with data from an t-file
# the precision and endian will be the same as in a-file
# this is in some way a python-version of the ext_mod script
# by Alberto, Ralf and Christof
# use the corresponding config-file to run this script
# usage:
# replace_codes.py <config-file>
# because of the need of numpy it does not work on blizzard so far
# (on lizard and all other mpi computers it works!)
#

import datetime
import ConfigParser
import logging
import shutil
import os
import sys

from optparse import OptionParser
from PyRemo.OoPlot.DataFile import IegFile

logging.basicConfig(level=logging.INFO,
                    format='%(asctime)s %(levelname)-8s %(message)s',
                    datefmt='%d %b %Y %H:%M:%S')

logging.info('Using Python version ' + sys.version)

parser = OptionParser("usage: %prog [options] <config_file.txt>")
(options, args) = parser.parse_args()

if len(args) != 1:
    parser.error('Wrong number of parameters!')
else:
    config_file = args[0]

# loading defaults
config = ConfigParser.SafeConfigParser()
config.readfp(open(config_file))

#
# Data specific configuration
#
afile_date = datetime.datetime(int(config.get('a-file', 'afile_year')),
                               int(config.get('a-file', 'afile_month')),
                               int(config.get('a-file', 'afile_day')),
                               int(config.get('a-file', 'afile_hour')))
rfile_date = datetime.datetime(int(config.get('replace-file', 'rfile_year')),
                               int(config.get('replace-file', 'rfile_month')),
                               int(config.get('replace-file', 'rfile_day')),
                               int(config.get('replace-file', 'rfile_hour')))

#
# Data paths
#
afile_path = config.get('paths', 'afile_path')
rfile_path = config.get('paths', 'rfile_path')

#
# Codes
#
codes_str = config.get('codes', 'code_list').split()
codes = []
for code in codes_str:
    codes.append(int(code))
logging.info("codes {0} will be replaced".format(codes))

#
# Read a-file and do a backup
#
afile_name = 'a{0}{1}a{2}'.format(config.get('a-file', 'afile_user'),
                                  config.get('a-file', 'afile_exp'),
                                  afile_date.strftime('%Y%m%d%H'))
afile_full_name = os.path.join(afile_path, afile_name)
afile_backup = '{0}.ori'.format(afile_full_name)
logging.info("backup of {0}".format(afile_full_name))

if os.path.isfile(afile_backup):
    logging.warning("backup file {0} already exists!".format(afile_backup))
else:
    shutil.move(afile_full_name, afile_backup)

afile_in = IegFile(afile_backup)
afile_endian = afile_in.is_big_endian()
afile_precision = afile_in.is_double_precision() 
afile_header, afile_arr, afile_cdic = afile_in.read_all_sel_data()
afile_in.close()

#
# Read replace-file
#

rfile_name = 'e{0}{1}{2}{3}'.format(config.get('replace-file', 'rfile_user'),
                                    config.get('replace-file', 'rfile_exp'),
                                    config.get('replace-file', 'rfile_type'),
                                    rfile_date.strftime('%Y%m%d%H'))
rfile_full_name = os.path.join(rfile_path, rfile_name)

rfile_in = IegFile(rfile_full_name)
rfile_header, rfile_arr, rfile_cdic = rfile_in.read_all_sel_data()
rfile_in.close()

#
# Write new a-file and take codes from codelist from replace-file
# the rest of the original a-file stays the same.
# The time information is taken from a-file.
#

with IegFile(afile_full_name, mode='w', big_endian=afile_endian,
             double_precision=afile_precision) as afile_out:
    for index, header in enumerate(rfile_header):
        if header.pdb[6] in codes:
            new_header = header
            new_header.pdb[10] = afile_date.year
            new_header.pdb[11] = afile_date.month
            new_header.pdb[12] = afile_date.day
            new_header.pdb[13] = afile_date.hour
            afile_out.write_data_set(new_header, rfile_arr[index])
            logging.info("using code {0} from replace-file".format(
                    rfile_header[index].pdb[6]))
    for index, header in enumerate(afile_header):
        if not header.pdb[6] in codes:
            afile_out.write_data_set(header, afile_arr[index])
            logging.info("using code {0} from a-file".format(
                    afile_header[index].pdb[6]))


autosubmit/jobs/ret_forcing.sh
# Returns the boundary files for the current year
#
set -ex
# needed at DKRZ to use modules from a script
source /sw/rhel6-x64/etc/profile.mistral
# load pftp module
module load pftp
#
module load python/2.7-ve0
#
function ftp_input {
cat > ftp.inp << EOF
cd $1/year$2
bin
get $3
quit
EOF
}

function download {
    #
    # *** style for achiving yearly|monthly
    YY=$1
    if [ "%BOUND_ARCH_STYLE%" = "monthly" ]
    then
	for MM in 01 02 03 04 05 06 07 08 09 10 11 12
	do
	    DNBOUND=a${BUSER}${BEXP}a${YY}${MM}.tar

	    ftp_input $CDIR $YY $DNBOUND

            pftp < ftp.inp
#
	done
#
    elif [ "%BOUND_ARCH_STYLE%" = "yearly" ]
    then
	DNBOUND=a${BUSER}${BEXP}a${YY}.tar

	ftp_input $CDIR $YY $DNBOUND
	
        pftp < ftp.inp
	tar xf a${BUSER}${BEXP}a${YY}.tar
	wait
	rm a${BUSER}${BEXP}a${YY}.tar
    else
	exit 2
    fi
}

YEAR=%Chunk_START_YEAR%
MONTH=%Chunk_START_MONTH%
BUSER=%BOUND_USER%         # for boundary data
BEXP=%BOUND_EXP%
CURRENT_FILE=a${BUSER}${BEXP}a${YEAR}${MONTH}.tar
#
###############################################################################
#
# Archive directory of boundary data
if [ %BOUND_EXP_STR_STYLE% = 1 ]
then
    CDIR=%BOUND_ARCH_PATH%/exp$BEXP
elif [ %BOUND_EXP_STR_STYLE% = 2 ]
then
    CDIR=%BOUND_ARCH_PATH%/exp${BUSER}${BEXP}
else
    exit 1
fi

PFADFRC=%HPCROOTDIR%/%MEMBER%/bound_tar  # local directory for boundary files

# Path for unpacked boundary files
if [ "%DATA_DIRS%" = "default" ]
then
    PFADXA=%HPCROOTDIR%/%MEMBER%/xa
else
    PFADXA=%PATH_A%
fi

cd $PFADFRC

# Download missing data
if [ ! -s $CURRENT_FILE ]
then
    download $YEAR
fi

# Download next year data
if [[ $MONTH = 12 && "%Chunk_LAST%" = "FALSE" ]]
then
    NEXT_YEAR=$(( $YEAR + 1 ))
    download $NEXT_YEAR
    
    # Delete boundary data of previous year
    if [[ "%Chunk_FIRST%" = "FALSE" && "%KEEP_BOUND%" = "False" ]]
    then
	PREV_YEAR=$(( $YEAR - 1 ))
	set +e
	rm a${BUSER}${BEXP}a${PREV_YEAR}??.tar
	set -ex
    fi
fi

# Empty boundary data directory
cd $PFADXA
set +e
rm a${BUSER}${BEXP}a*
set -ex

# Copy and unpack current file to boundary file path
cp $PFADFRC/$CURRENT_FILE .
tar xf $CURRENT_FILE
wait
rm $CURRENT_FILE

# In case of warmstart replace soilparameters
if [[ "%Chunk_FIRST%" = "TRUE" && "%WARMSTART%" = "True" ]]
then
    WARM_CONF=%HPCROOTDIR%/%MEMBER%/warmstart.conf
    REPLACE_SCRIPT=%HPCROOTDIR%/REMO_MPI/autosubmit/jobs/replace_codes.py
    python $REPLACE_SCRIPT $WARM_CONF
fi
